File: examples/widgets/controls/TextBox.m
#include <Cocoa/Cocoa.h>

static const char* nsstr(NSString* s) { return s ? [s UTF8String] : ""; }

@interface Window : NSWindow {
  NSTextField* textBox1;
  NSTextField* textBox2;
}
- (instancetype)init;
- (BOOL)windowShouldClose:(id)sender;
- (void)dumpLayout:(NSString*)stage;
@end

@implementation Window
- (void)dumpView:(NSView*)view name:(NSString*)name {
  NSRect frame = [view frame];
  NSRect bounds = [view bounds];
  NSLog(@"[%@] frame=(%.1f,%.1f %.1fx%.1f) bounds=(%.1f,%.1f %.1fx%.1f) autoresizeMask=0x%lx",
      name,
      frame.origin.x, frame.origin.y, frame.size.width, frame.size.height,
      bounds.origin.x, bounds.origin.y, bounds.size.width, bounds.size.height,
      (unsigned long)[view autoresizingMask]);
  printf("[%s] frame=(%.1f,%.1f %.1fx%.1f) bounds=(%.1f,%.1f %.1fx%.1f) autoresizeMask=0x%lx\n",
      nsstr(name),
      frame.origin.x, frame.origin.y, frame.size.width, frame.size.height,
      bounds.origin.x, bounds.origin.y, bounds.size.width, bounds.size.height,
      (unsigned long)[view autoresizingMask]);
}

- (void)dumpLayout:(NSString*)stage {
  NSRect windowFrame = [self frame];
  NSRect contentRect = [self contentRectForFrameRect:windowFrame];
  NSLog(@"[Window %@] frame=(%.1f,%.1f %.1fx%.1f) contentRect=(%.1f,%.1f %.1fx%.1f)",
      stage,
      windowFrame.origin.x, windowFrame.origin.y, windowFrame.size.width, windowFrame.size.height,
      contentRect.origin.x, contentRect.origin.y, contentRect.size.width, contentRect.size.height);
  printf("[Window %s] frame=(%.1f,%.1f %.1fx%.1f) contentRect=(%.1f,%.1f %.1fx%.1f)\n",
      nsstr(stage),
      windowFrame.origin.x, windowFrame.origin.y, windowFrame.size.width, windowFrame.size.height,
      contentRect.origin.x, contentRect.origin.y, contentRect.size.width, contentRect.size.height);
  [self dumpView:[self contentView] name:@"contentView"];
  [self dumpView:textBox1 name:@"textBox1"];
  [self dumpView:textBox2 name:@"textBox2"];
  NSLog(@"[textBox1] string='%@' editable=%d selectable=%d bezeled=%d drawsBackground=%d",
      [textBox1 stringValue], [textBox1 isEditable], [textBox1 isSelectable], [textBox1 isBezeled], [textBox1 drawsBackground]);
  NSLog(@"[textBox2] string='%@' editable=%d selectable=%d bezeled=%d drawsBackground=%d firstResponder=%d",
      [textBox2 stringValue], [textBox2 isEditable], [textBox2 isSelectable], [textBox2 isBezeled], [textBox2 drawsBackground],
      [[self firstResponder] isEqual:textBox2]);
}

- (instancetype)init {
  textBox1 = [[[NSTextField alloc] initWithFrame:NSMakeRect(10, 270, 100, 20)] autorelease];
  [textBox1 setStringValue:@"textBox1"];
  
  textBox2 = [[[NSTextField alloc] initWithFrame:NSMakeRect(10, 230, 100, 20)] autorelease];
  [textBox2 setStringValue:@"textBox2"];
  
  [super initWithContentRect:NSMakeRect(100, 100, 300, 300) styleMask:NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable | NSWindowStyleMaskResizable backing:NSBackingStoreBuffered defer:NO];
  [self setTitle:@"TextBox Example"];
  [[self contentView] addSubview:textBox1];
  [[self contentView] addSubview:textBox2];
  [self makeFirstResponder:textBox2];
  [self setIsVisible:YES];
  [self dumpLayout:@"init"];
  return self;
}

- (BOOL)windowShouldClose:(id)sender {
  [NSApp terminate:sender];
  return YES;
}
@end

int main(int argc, char *argv[]) {
    [NSApplication sharedApplication];
    [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];

    Window *window = [[Window alloc] init];
    [window makeKeyAndOrderFront:nil];

    [NSApp activateIgnoringOtherApps:YES];
    if (getenv("TEXTBOX_DUMP_LAYOUT_ONCE") != NULL) {
      [NSApp performSelector:@selector(terminate:) withObject:nil afterDelay:0.2];
    }
    [NSApp run];
    return 0;
}

File: examples/widgets/controls/textbox.nim
import std/[os, strutils]

import merenda/appkit
import merenda/objc

proc maxFramesFromEnv(defaultValue = -1): int =
  let raw = getEnv("MERENDA_EXAMPLE_FRAMES").strip()
  if raw.len == 0:
    return defaultValue
  try:
    parseInt(raw)
  except ValueError:
    defaultValue

proc debugRenderDumpEnabled(): bool =
  getEnv("MERENDA_APPKIT_DEBUG_RENDER").strip().toLowerAscii() in
    ["1", "true", "yes", "on"]

proc format1(value: float32): string =
  formatFloat(value.float, ffDecimal, 1)

proc formatRect(rect: NSRect): string =
  "(" & format1(rect.origin.x) & "," & format1(rect.origin.y) & " " &
    format1(rect.size.width) & "x" & format1(rect.size.height) & ")"

proc intBool(flag: bool): int =
  if flag: 1 else: 0

proc dumpViewLine(view: NSView, name: string) =
  if view.isNil:
    return
  echo "[",
    name,
    "] frame=",
    formatRect(view.frame()),
    " bounds=",
    formatRect(view.bounds()),
    " autoresizeMask=0x",
    toHex(view.autoresizingMask(), 2)

proc dumpTextBoxLayout(
    stage: string, window: NSWindow, textBox1: NSTextField, textBox2: NSTextField
) =
  if window.isNil:
    return
  let windowFrame = window.frame()
  let contentRect = window.contentRectForFrameRect(windowFrame)
  let contentView = window.contentView()
  let firstResponder = window.firstResponder()
  echo "[Window ",
    stage, "] frame=", formatRect(windowFrame), " contentRect=", formatRect(contentRect)
  dumpViewLine(contentView, "contentView")
  dumpViewLine(textBox1.NSView, "textBox1")
  dumpViewLine(textBox2.NSView, "textBox2")
  echo "[textBox1] string='",
    textBox1.stringValue(),
    "' editable=",
    intBool(textBox1.isEditable()),
    " selectable=",
    intBool(textBox1.isSelectable()),
    " bezeled=",
    intBool(textBox1.isBezeled()),
    " drawsBackground=",
    intBool(textBox1.drawsBackground())
  echo "[textBox2] string='",
    textBox2.stringValue(),
    "' editable=",
    intBool(textBox2.isEditable()),
    " selectable=",
    intBool(textBox2.isSelectable()),
    " bezeled=",
    intBool(textBox2.isBezeled()),
    " drawsBackground=",
    intBool(textBox2.drawsBackground()),
    " firstResponder=",
    intBool((not firstResponder.isNil) and firstResponder.value == textBox2.value)
  if debugRenderDumpEnabled():
    debugDumpWindowRenderTree(window)

objcImpl:
  type TextBoxWindow = object of NSWindow
    xTextBox1: NSTextField
    xTextBox2: NSTextField

  method init*(self: var TextBoxWindow): TextBoxWindow =
    var base = self.initWithContentRect(
      100.0,
      100.0,
      300.0,
      328.0,
      {NSTitledWindow, NSClosableWindow, NSMiniaturizableWindow, NSResizableWindow},
      NSBackingStoreBuffered,
      false,
    )
    result = asTypeRaw[TextBoxWindow](base.value)
    base.value = nil
    if result.isNil:
      return

    var contentAlloc = NSView.alloc()
    var contentView = contentAlloc.initWithFrame(nsRect(0.0, 0.0, 300.0, 300.0))
    contentAlloc.value = nil
    result.setContentView(contentView)

    var textBox1Alloc = NSTextField.alloc()
    result.xTextBox1 = textBox1Alloc.init()
    result.xTextBox1.setFrame(nsRect(10.0, 270.0, 100.0, 20.0))
    result.xTextBox1.setStringValue(@ns"textBox1")
    result.xTextBox1.setAlignment(NSLeftTextAlignment)

    var textBox2Alloc = NSTextField.alloc()
    result.xTextBox2 = textBox2Alloc.init()
    result.xTextBox2.setFrame(nsRect(10.0, 230.0, 100.0, 20.0))
    result.xTextBox2.setStringValue(@ns"textBox2")
    result.xTextBox2.setAlignment(NSLeftTextAlignment)

    contentView.addSubview(result.xTextBox1)
    contentView.addSubview(result.xTextBox2)
    discard result.makeFirstResponder(result.xTextBox2.NSResponder)
    result.setTitle(@ns"TextBox Example")
    result.setIsVisible(true)
    dumpTextBoxLayout("init", result as NSWindow, result.xTextBox1, result.xTextBox2)
    contentView.value = nil

  method windowShouldClose*(self: TextBoxWindow, sender: NSObject): bool =
    if sender.isNil:
      NSApp().stop()
      return true
    NSApp().stop()
    true

  method dealloc(self: TextBoxWindow) {.used.} =
    self.xTextBox1 = NSTextField(value: nil)
    self.xTextBox2 = NSTextField(value: nil)
    destroyIvarFields(self)
    discard callSuperIdFrom(TextBoxWindow, self, getSelector("dealloc"))

when isMainModule:
  var app = NSApp()
  var windowAlloc = TextBoxWindow.alloc()
  var window = initOwned(move(windowAlloc))

  app.addWindow(window.NSWindow)
  window.makeKeyAndOrderFront(app.NSObject)

  try:
    let maxFrames = maxFramesFromEnv()
    if maxFrames < 0:
      app.run()
    else:
      discard app.runForFrames(maxFrames)
  except CatchableError as exc:
    echo "Unable to run textbox example: ", exc.msg

  window.value = nil
  app.value = nil
